tests/rpc/test_rpc_eth_blocks.py

```diff
@@ -0,0 +1,136 @@
+import json
+import os
+import pytest
+
+from eth_utils import (
+    decode_hex,
+)
+
+from evm import MainnetTesterChain
+from evm.db.backends.memory import MemoryDB
+from evm.db.chain import BaseChainDB
+from evm.rpc import RPCServer
+
+
+@pytest.fixture
+def chain():
+    server_test_path = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+    db_path = os.path.join(server_test_path, 'fixtures', 'rpc_test_chain.db')
+    db = MemoryDB()
+    with open(db_path) as f:
+        key_val_hex = json.loads(f.read())
+        db.kv_store = {decode_hex(k): decode_hex(v) for k, v in key_val_hex.items()}
```

Comment:
Does this mean we'll have to regenerate `rpc_test_chain.db` every time there's a change to the DB schema? Did you consider storing just the RLP data for the blocks and then here decoding/inserting them in the DB using the public APIs, like in https://github.com/ethereum/py-evm/blob/master/evm/p2p/test_peer.py#L452?

Reply:
How often is the schema changing? I assume that at some point the schema will be locked down, or that the database would have to handle its own schema upgrades. Otherwise, people's stored chain db could become corrupt at each `py-evm` upgrade.

I just tried loading that data set you linked to, and `chain.get_canonical_block_by_number(...)` works with 0, 1, and 2, but then on 3, I get an error. Any idea if this is a limitation of the approach or a corrupt database?

```
~/code/py-evm/evm/chains/chain.py in get_canonical_block_by_number(self, block_number)
    165         """
    166         validate_uint256(block_number, title="Block Number")
--> 167         return self.get_block_by_hash(self.chaindb.lookup_block_hash(block_number))
    168 
    169     def get_block_by_hash(self, block_hash):

~/code/py-evm/evm/chains/chain.py in get_block_by_hash(self, block_hash)
    173         validate_word(block_hash, title="Block Hash")
    174         block_header = self.get_block_header_by_hash(block_hash)
--> 175         vm = self.get_vm(block_header)
    176         return vm.get_block_by_header(block_header)
    177 

~/code/py-evm/evm/chains/chain.py in get_vm(self, header)
    135 
    136         vm_class = self.get_vm_class_for_block_number(header.block_number)
--> 137         return vm_class(header=header, chaindb=self.chaindb)
    138 
    139     #

~/code/py-evm/evm/vm/base.py in __init__(self, header, chaindb)
     40         self.chaindb = chaindb
     41         block_class = self.get_block_class()
---> 42         self.block = block_class.from_header(header=header, chaindb=self.chaindb)
     43 
     44     @classmethod

~/code/py-evm/evm/vm/forks/frontier/blocks.py in from_header(cls, header, chaindb)
    247             uncles = []
    248         else:
--> 249             uncles = chaindb.get_block_uncles(header.uncles_hash)
    250 
    251         transactions = chaindb.get_block_transactions(header, cls.get_transaction_class())

~/code/py-evm/evm/db/chain.py in get_block_uncles(self, uncles_hash)
    143     def get_block_uncles(self, uncles_hash):
    144         validate_word(uncles_hash, title="Uncles Hash")
--> 145         return rlp.decode(self.db.get(uncles_hash), sedes=rlp.sedes.CountableList(BlockHeader))
    146 
    147     @to_list

~/code/py-evm/evm/db/journal.py in get(self, key)
    135 
    136     def get(self, key):
--> 137         return self.wrapped_db.get(key)
    138 
    139     def set(self, key, value):

~/code/py-evm/evm/db/backends/memory.py in get(self, key)
     11 
     12     def get(self, key):
---> 13         return self.kv_store[key]
     14 
     15     def set(self, key, value):

KeyError: b"k\x17\xb98\xc6\xe4\xef\x18\xb2j\xd8\x1b\x9c\xa3Q_'\xfd\x9cN\x82\xaa\xc5j\x1f\xd8\xea\xb2\x88x^A"
```

Reply:
At this early stage I wouldn't be surprised if we changed the db schema frequently. We'd probably version those schemas so that existing user chain data can continue to be used and/or migrated, but ideally we shouldn't have to worry about that in those tests and that's why I suggested an approach that is not locked to a specific db schema.

That data set I linked to contains only the header data (the README describes exactly what is there), so I'm surprised you were even able to use get_canonical_block_by_number() with it